### PR TITLE
Optimize comboTp

### PR DESCRIPTION
**Pull request by joseph-ireland**
1. optimizes calculateComboTps 
2. adds an approximation of math.pow(2,x) used in a core loop

overall 2-3x full program speed improvement

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs [hunk 1]
```diff
@@ -233,73 +233,31 @@ private static double getCheeseNoteCount(IList<OsuMovement> movements, double tp
             return count;
         }
 
-        private static double[] calculateComboTps(List<OsuMovement> movements)
+        private static double[] calculateComboTps(HitProbabilities movements)
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs]: its not really `movements` anymore

File: osu.Game.Rulesets.Osu/Difficulty/MathUtil/HitProbabilities.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using osu.Game.Rulesets.Osu.Difficulty.Preprocessing;
+
+namespace osu.Game.Rulesets.Osu.Difficulty.MathUtil
+{
+    class HitProbabilities
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/MathUtil/HitProbabilities.cs]: access modifier

File: osu.Game.Rulesets.Osu/Difficulty/MathUtil/HitProbabilities.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        {
+            int count = 0;
+            for (int i = start; i != start + sectionCount; i++)
+            {
+                count += sections[i].Movements.Count;
+            }
+
+            return count;
+        }
+        public bool IsNull(int sectionCount)
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/MathUtil/HitProbabilities.cs]: `IsEmpty` would be better imo, `IsNull` sounds like its checking for `null`

Comment (joseph-ireland) [reply osu.Game.Rulesets.Osu/Difficulty/MathUtil/HitProbabilities.cs]: agree with everything

**Pull request merged**